chain/epoch_manager/src/lib.rs

```diff
@@ -519,35 +525,42 @@ impl EpochManager {
         &mut self,
         epoch_id: &EpochId,
         last_known_block_hash: &CryptoHash,
-    ) -> Result<Vec<(ValidatorStake, bool)>, EpochError> {
-        let slashed = self.get_slashed_validators(last_known_block_hash)?.clone();
-        let epoch_info = self.get_epoch_info(epoch_id)?;
-        let mut settlement = vec![];
-        for validator_id in epoch_info.block_producers_settlement.iter() {
-            let validator_stake = epoch_info.validators[*validator_id as usize].clone();
-            let is_slashed = slashed.contains_key(&validator_stake.account_id);
-            settlement.push((validator_stake, is_slashed));
+    ) -> Result<&[(ValidatorStake, bool)], EpochError> {
+        // TODO: Revisit this when we enable slashing
+        if self.epoch_validators_ordered.cache_get(epoch_id).is_none() {
```

Comment:
Could maybe optimize for the case where the `epoch_id` is known by using a `match` (presently there are always two cache lookups, even when already present)

```rust
match self.epoch_validators_ordered.cache_get(epoch_id) {
    Some(settlement) => Ok(settlement),
    None => {
        ... // compute settlement
        self.epoch_validators_ordered.cache_set(epoch_id.clone(), settlement);
        Ok(self.epoch_validators_ordered.cache_get(epoch_id).unwrap())
    }
}
```

Reply:
cannot get around lifetime restrictions

Reply:
I see. I guess since the function returns a reference it needs to thread the same lifetime all the way through no matter how you scope it. Ok, I suppose we have to live with 2 cache hits per call. Thanks for checking.